### PR TITLE
Update file and define conventions

### PR DESCRIPTION
**Pull request by weshaggard**
@roncain @CIPop @davidsh @stephentoub please have a look.

### CODE REVIEW

File: Documentation/coding-guidelines/project-guidelines.md [hunk 1]
```diff
@@ -97,20 +97,27 @@ Facade projects are unique in that they don't have any code and instead are gene
...
-
-Where `<group>` is:
-- Windows/Linux/Unix/Osx - For files specific to a given OS
-- WinRT - For files that are have unique WinRT dependencies
-- Target Runtime
- - NetFx - Files specific for the full .NET Framework/CLR runtime
- - CoreCLR - Files specific to the CoreCLR runtime
- - NETNative - Files specific to the .NET Native runtime (MRT) or toolchain.
+# Conventions for forked code
+While our goal is to have the exact code for every configuration there is always reasons why that is not realistic so we need to have a set of conventions for dealing with places were we fork code. In order of preference, here are the strategies we employ:
```
Comment (stephentoub) [inline Documentation/coding-guidelines/project-guidelines.md]: Nit: were => where

File: Documentation/coding-guidelines/project-guidelines.md [hunk 2]
```diff
@@ -97,20 +97,27 @@ Facade projects are unique in that they don't have any code and instead are gene
...
+
+## Code file naming conventions
+Each source file should use the following guidelines
+- The source code file should contain only one class. The only exception is small supporting structs, enums, nested classes, or delegates that only apply to the class can also be contained in the source file. 
+- The source code file should be named `<class>.cs` and should be placed in a directory structure that matches its namespace relative to its project directory. Ex. `System\IO\Stream.cs` 
+- Larger nested classes should be factored out into their own source files using a partial class and the file name should be `<class>.<nested class>.cs`.
+- Classes that are forked based on configuration should have file names `<class>.<configuration>.cs`. 
+ - Where `<configuration>` is one of `$(OSGroup)`, `$(TargetGroup)`, `$(ConfigurationGroup)`, or `$(Platform)`.
+- Classes that are forked based on a feature set should have file names `<class>.<feature>.cs`.
+ - Where `<feature>`is the name something that causes a fork in code that isn't a configuration such as WinRT support. Ex. `FileSystemInfo..WinRT.cs`.
```
Comment (stephentoub) [inline Documentation/coding-guidelines/project-guidelines.md]: Nit: extra period in the filename
Nit: "is the name something" => "is the name of something"

**Comment by stephentoub** [comment]
LGTM

**Comment by davidsh** [comment]
@weshaggard 

For $(TargetGroup) such as "netcore50", "dnxcore50", etc. are you proposing to use lower or upper case letters?  Usually ifdef's are all uppercase. Not sure about filenames.

So, should we do this in code:

```c#
#ifdef NETCORE50
// code
#endif
```

or

```c#
#ifdef netcore50
// code
#endif

And what about filenames, i.e. 

`HttpClientHandler.NETCORE50.cs`

or

`HttpClientHandler.netcore50.cs`

**Comment by weshaggard** [comment]
Thanks @stephentoub I made those minor edits.

@davidsh I added some clarification on casing. For consistency I think it is best to match the exact case everywhere, and for the targets they are usually all lowercase. I realize that isn't the norm for defines but I think consistency out weights that norm for me.

**Pull request merged**